File: 37-sudoku-solver/sudoku-solver.py
class Solution:
    def solveSudoku(self, board):
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxes = [set() for _ in range(9)]

        for i in range(9):
            for j in range(9):
                if board[i][j] != '.':
                    rows[i].add(board[i][j])
                    cols[j].add(board[i][j])
                    boxes[(i // 3) * 3 + (j // 3)].add(board[i][j])

        def backtrack():
            for i in range(9):
                for j in range(9):
                    if board[i][j] == '.':
                        box_idx = (i // 3) * 3 + (j // 3)
                        for ch in '123456789':
                            if (
                                ch not in rows[i]
                                and ch not in cols[j]
                                and ch not in boxes[box_idx]
                            ):
                                board[i][j] = ch
                                rows[i].add(ch)
                                cols[j].add(ch)
                                boxes[box_idx].add(ch)

                                if backtrack():
                                    return True

                                board[i][j] = '.'
                                rows[i].remove(ch)
                                cols[j].remove(ch)
                                boxes[box_idx].remove(ch)
                        return False
            return True

        backtrack()

File: 37-sudoku-solver/sudoku-solver.js
var solveSudoku = function (board) {
    const rows = Array.from({ length: 9 }, () => new Set());
    const cols = Array.from({ length: 9 }, () => new Set());
    const boxes = Array.from({ length: 9 }, () => new Set());

    for (let i = 0; i < 9; i++) {
        for (let j = 0; j < 9; j++) {
            if (board[i][j] !== '.') {
                rows[i].add(board[i][j]);
                cols[j].add(board[i][j]);
                boxes[Math.floor(i / 3) * 3 + Math.floor(j / 3)].add(board[i][j]);
            }
        }
    }

    function backtrack() {
        for (let i = 0; i < 9; i++) {
            for (let j = 0; j < 9; j++) {
                if (board[i][j] === '.') {
                    let boxIndex = Math.floor(i / 3) * 3 + Math.floor(j / 3);

                    for (let ch = 1; ch <= 9; ch++) {
                        let num = ch.toString();
                        if (
                            !rows[i].has(num) &&
                            !cols[j].has(num) &&
                            !boxes[boxIndex].has(num)
                        ) {
                            board[i][j] = num;
                            rows[i].add(num);
                            cols[j].add(num);
                            boxes[boxIndex].add(num);

                            if (backtrack()) return true;

                            board[i][j] = '.';
                            rows[i].delete(num);
                            cols[j].delete(num);
                            boxes[boxIndex].delete(num);
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }

    backtrack();
};
